File: code_samples/section9/problems/problem9_3/problem9_3.java
package code_samples.section9.problems.problem9_3;

/**
 * problem9_3
 * ----------
 * Solves the "Partition Equal Subset Sum" problem using dynamic programming.
 *
 * Given an array of positive integers, this class determines whether the array
 * can be partitioned into two subsets such that the sum of elements in both
 * subsets is equal.
 */
public class problem9_3 {

    /**
     * Determines whether the given array can be partitioned into two subsets
     * with equal sums.
     *
     * Key observation:
     *   - Let S be the sum of all elements.
     *   - If S is odd, it is impossible to split it into two equal integer sums.
     *   - If S is even, the problem reduces to checking whether there exists
     *     a subset whose sum is exactly S / 2.
     *
     * Dynamic Programming formulation (1D boolean DP):
     *   dp[s] = true if a subset of the processed elements can form sum s.
     *
     * Initialization:
     *   dp[0] = true
     *   (A sum of 0 is always achievable by choosing no elements.)
     *
     * Transition:
     *   For each element x:
     *     for s from target down to x:
     *       if dp[s - x] is true, then dp[s] can be set to true.
     *
     * Important detail:
     *   - The inner loop iterates backward to ensure each element is used
     *     at most once (0/1 subset-sum).
     *   - Iterating forward would allow reusing the same element multiple times.
     *
     * Time Complexity:
     *   O(n * target), where target = totalSum / 2
     *
     * Space Complexity:
     *   O(target)
     *
     * @param nums Input array of positive integers
     * @return true if the array can be partitioned into two equal-sum subsets,
     *         false otherwise
     */
    boolean canPartition(int[] nums) {
        // Compute the total sum of all elements
        int total = 0;
        for (int x : nums) total += x;

        // If total sum is odd, equal partition is impossible
        if ((total & 1) == 1) return false;

        // Target sum for one subset
        int target = total / 2;

        // dp[s] indicates whether sum s is achievable
        boolean[] dp = new boolean[target + 1];

        // Base case: sum 0 is always achievable
        dp[0] = true;

        // Process each number in the array
        for (int x : nums) {
            // Traverse sums backwards to enforce 0/1 usage of each element
            for (int s = target; s >= x; s--) {
                // If we could previously make (s - x), then we can make s by adding x
                if (dp[s - x]) dp[s] = true;
            }
        }

        // If target sum is achievable, a valid partition exists
        return dp[target];
    }

    // ===========================
    // Test Harness
    // ===========================

    /**
     * Entry point for running predefined test cases.
     *
     * Each test prints:
     *   - Test name
     *   - Input array
     *   - Computed result
     *   - Expected result
     */
    public static void main(String[] args) {
        problem9_3 solver = new problem9_3();

        runTest("Test 1: Classic example", new int[]{1, 5, 11, 5}, true, solver);
        runTest("Test 2: Cannot partition", new int[]{1, 2, 3, 5}, false, solver);
        runTest("Test 3: Single element", new int[]{7}, false, solver);
        runTest("Test 4: Two equal numbers", new int[]{4, 4}, true, solver);
        runTest("Test 5: Larger mixed set", new int[]{2, 2, 3, 5}, false, solver);
        runTest("Test 6: Multiple valid subsets", new int[]{2, 2, 1, 1}, true, solver);
    }

    /**
     * Helper method to run and display a single test case.
     *
     * @param name      Descriptive test name
     * @param arr       Input array
     * @param expected  Expected boolean result
     * @param solver    Instance of problem9_3 used to execute the algorithm
     */
    private static void runTest(String name, int[] arr, boolean expected, problem9_3 solver) {
        // Run the partition check
        boolean result = solver.canPartition(arr);

        // Print test details
        System.out.println(name);
        System.out.print("Input: {");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i + 1 < arr.length) System.out.print(",");
        }
        System.out.println("}");

        // Print result versus expected outcome
        System.out.println("canPartition = " + result + " (expected " + expected + ")\n");
    }
}

File: code_samples/section9/problems/problem9_3/problem9_3.py
"""Problem 9.3 - Partition Equal Subset Sum."""

def can_partition(nums: list[int]) -> bool:
    """Return True if nums can be partitioned into two subsets with equal sum.

    This function solves the classic *Partition Equal Subset Sum* problem,
    which can be reduced to a 0/1 subset-sum decision problem.

    Key idea:
        - Let S be the sum of all elements.
        - If S is odd, it is impossible to split the array into two subsets
          with equal integer sums.
        - If S is even, the problem becomes:
              "Is there a subset of nums that sums to S // 2?"

    Dynamic Programming formulation (1D boolean DP):
        dp[s] = True if there exists a subset of the processed elements
                whose sum is exactly s.

    Initialization:
        dp[0] = True
        (A sum of 0 is always achievable by choosing no elements.)

    Transition:
        For each element x in nums:
            for s from target down to x:
                if dp[s - x] is True:
                    dp[s] = True

    Important detail:
        - The inner loop iterates *backwards* to ensure each element is used
          at most once (0/1 choice).
        - Iterating forwards would allow the same element to be reused,
          turning this into an unbounded subset-sum problem.

    Complexity:
        Time:  O(n * target), where target = sum(nums) // 2
        Space: O(target)

    Args:
        nums: List of positive integers.

    Returns:
        True if nums can be partitioned into two subsets with equal sum,
        False otherwise.
    """
    # Compute the total sum of all elements.
    total = sum(nums)

    # If the total sum is odd, it cannot be split into two equal halves.
    if total % 2 == 1:
        return False

    # Target sum for one subset.
    target = total // 2

    # dp[s] indicates whether sum s is achievable.
    dp = [False] * (target + 1)

    # Base case: sum 0 is always achievable.
    dp[0] = True

    # Process each number in the array.
    for x in nums:
        # Traverse sums backwards to enforce 0/1 usage of each element.
        for s in range(target, x - 1, -1):
            # If we could previously make (s - x),
            # then we can make s by including x.
            if dp[s - x]:
                dp[s] = True

    # If target sum is achievable, a valid partition exists.
    return dp[target]

# ===========================
# Test Harness
# ===========================

def test(name, arr, expected):
    """Test can_partition.

    Prints:
        - Test name
        - Input array
        - Computed result
        - Expected result
    """
    result = can_partition(arr)
    print(name)
    print(f"Input: {arr}")
    print(f"can_partition = {result} (expected {expected})\n")


# Run a series of sanity checks covering common cases.
print("=== Testing can_partition ===\n")

# Classic example: partition exists (sum = 22, target = 11).
test("Test 1: Classic example", [1, 5, 11, 5], True)

# Total sum is odd => cannot partition.
test("Test 2: Cannot partition", [1, 2, 3, 5], False)

# Single element cannot be split into two equal-sum subsets.
test("Test 3: Single element", [7], False)

# Two equal numbers can be partitioned trivially.
test("Test 4: Two equal numbers", [4, 4], True)

# Even total sum, but no subset achieves target.
test("Test 5: Larger mixed set", [2, 2, 3, 5], False)

# Multiple valid subsets exist (e.g., {2,1} and {2,1}).
test("Test 6: Multiple valid subsets", [2, 2, 1, 1], True)
